### ISSUE TITLE
Add a minimal Dockerfile

### ISSUE THREAD

**Issue by kargakis** [mimblewimble/grin#919]
I am suggesting to move the base-build Dockerfile from [grin-docker](https://github.com/mimblewimble/grin-docker/blob/master/base-build/Dockerfile) inside this repo. It will be easier to keep it up to date and track grin's build dependencies. Even better build an image out of it as part of the CI testing of this repo.

**Comment by sesam** [comment]
How would that affect TravisCI build times? We've been having trouble getting the CI to work for, rather than against, us.

Related:
https://github.com/mimblewimble/grin/issues/749
https://github.com/mimblewimble/grin/issues/772
https://github.com/mimblewimble/grin/issues/849
https://github.com/mimblewimble/grin/issues/190

**Comment by kargakis** [comment]
The docker build could be a separate job if it ends up slowing down the main test suite. Unclear whether you can have more than one job with travis though. How is the gitlab work going? Maybe evaluate other options for extra jobs like CircleCI?

**Comment by ignopeverell** [comment]
I agree with the suggestion. We can let people build the image themselves for now but a base Dockerfile could be handy for many.

**Comment by yeastplume** [comment]
I don't think I agree with moving the docker file into the main Grin repo.. it seems a widely-adopted practice to keep docker builds in separate repositories, which makes sense to me as it's very much a layer on top of the core build. Getting people to use it manually is just a documentation issue, we just need to tell people that it's there and tell them how to build it.

The reason it keeps falling out of date is that no automated build is using it and there's no automated testing on it... I'd be in favour of having a nightly docker build that at least checks that it builds and puts the latest code onto docker hub.

**Comment by kargakis** [comment]
> I don't think I agree with moving the docker file into the main Grin repo.. it seems a widely-adopted practice to keep docker builds in separate repositories, which makes sense to me as it's very much a layer on top of the core build.

I am working on production environments that base builds/deployments off of docker images and I can ensure you this is not true. Even though production images are using package managers to do installs, I don't think you are worried about that. There are two different modes optimized for development; one being bind-mounting local repos to simulate production installs and another one being bind-mounting the binaries directly (fastest). In both cases, dockerfiles are part of your source code. Rule of thumb: keeping build/deployment artifacts alongside your code is a best practice (spec files, dockerfiles, config files, what have you).

**Comment by kargakis** [comment]
> The reason it keeps falling out of date is that no automated build is using it and there's no automated testing on it... I'd be in favour of having a nightly docker build that at least checks that it builds and puts the latest code onto docker hub.

I don't see anybody stepping up to fix what you are asking here; a CI system that triggers builds on a different repo for every commit that lands in grin. It's a complicated thing to do anyway. Keeping the dockerfile in this repo is much simpler.

**Comment by yeastplume** [comment]
From what I can gather, every single one of the top projects listed in docker hub seem to be storing their docker builds in a separate repository somewhere: https://hub.docker.com/explore/ (whether it's in the official docker-library or their own). I'm sure whatever your company is doing works for them, but are these based on any well-known or general guidelines for how best to store and maintain docker images?

Also, I don't think it's a massive challenge to create a separate docker build in travis CI, which might be even easier given it's a separate repo, (and I don't want to add the mess of a docker build and deployment to the CI build on every grin commit, the CI process already takes far too long). There are details on how to do so here: https://docs.travis-ci.com/user/docker/ I'm sure there's a way to trigger it nightly, and will that not achieve the goal of having an up-to-date docker image with minimal messing?

**Comment by kargakis** [comment]
> From what I can gather, every single one of the top projects listed in docker hub seem to be storing their docker builds in a separate repository somewhere: https://hub.docker.com/explore/ (whether it's in the official docker-library or their own). I'm sure whatever your company is doing works for them, but are these based on any well-known or general guidelines for how best to store and maintain docker images?

I don't think there are any official guidlines. Most projects predate the container bubble, so this seems natural. For me as a developer, it's much easier to work on a single repo and not need to jump through hoops in order to build an image.

> and I don't want to add the mess of a docker build and deployment to the CI build on every grin commit, the CI process already takes far too long

You should be able to run a separate job for testing the docker image. 

> I'm sure there's a way to trigger it nightly, and will that not achieve the goal of having an up-to-date docker image with minimal messing?

It saves time to know which exact commit broke something than git bisecting your way out of it. Unless you are going to test against every commit, this will be a problem (maybe not a big one for the grin community). Even better, do not merge a PR until all jobs are passing so you guarantee you never have a broken thingy on master (which is huge).

**Comment by yeastplume** [comment]
I can see your points, but I don't think it's crucial that the docker build be up-to-the minute verified on each Grin commit at the moment. I'd see it as a deployment option similar to packages for each distro (albeit one that can be extended to testing); we'll definitely want to have packages for the major linux distributions, but I definitely don't think want each and every one in the main codebase.

If we manage to build out a docker testing environment, (i.e. a simulated network that gets rebuilt fairly frequently,) then we'll definitely need to consider something more ornate. But if it's just a quality-of-life package for the time being for those who want docker images, nightly should suffice.

**Comment by kargakis** [comment]
> I'd see it as a deployment option similar to packages for each distro (albeit one that can be extended to testing); we'll definitely want to have packages for the major linux distributions, but I definitely don't think want each and every one in the main codebase.

Why? No one is doing this. Just pick a base image and roll with it. Still, if you end up templatizing your Dockerfile for whatever reason just keep it in its own directory in this repo. Not a big deal.

**Comment by yeastplume** [comment]
I meant we'll have actual packages for debian, centos, arch, etc, (not separate docker packages with different bases), and 'docker' is just another distribution package in this regard. Yes, the docker package can use whatever base image is most appropriate.

**Comment by kargakis** [comment]
Right, misunderstood. In any case, making grin self-contained seems more developer-friendly.
